ipfs/client.go

```diff
@@ -0,0 +1,167 @@
+func StatCID(ipfsAPI string, cid string) (info *FileInfo, retErr error) {
+	if ipfsAPI == "" {
+		return nil, fmt.Errorf("specify IPFS API address")
+	}
+	ipfsAPIFilesStat := ipfsAPI + "/api/v0/files/stat"
+	req, err := http.NewRequest("POST", ipfsAPIFilesStat, nil)
+	if err != nil {
+		return nil, err
+	}
+	q := req.URL.Query()
+	q.Add("arg", "/ipfs/"+cid)
+	req.URL.RawQuery = q.Encode()
+	client := http.DefaultClient
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		if retErr != nil {
+			io.Copy(io.Discard, resp.Body)
+			resp.Body.Close()
+		}
+	}()
+	if resp.StatusCode != 200 {
+		return nil, fmt.Errorf("failed to stat %v; status code: %v", cid, resp.StatusCode)
+	}
+	var rs FileInfo
+	if err := json.NewDecoder(resp.Body).Decode(&rs); err != nil {
+		return nil, err
+	}
+	return &rs, nil
+}
+
+// Cat get the reader of the data specified by the IPFS path and optionally with
+// the offset and length.
+func Cat(ipfsAPI string, p string, offset *int, length *int) (_ io.ReadCloser, retErr error) {
+	if ipfsAPI == "" {
+		return nil, fmt.Errorf("specify IPFS API address")
+	}
+	ipfsAPICat := ipfsAPI + "/api/v0/cat"
+	req, err := http.NewRequest("POST", ipfsAPICat, nil)
+	if err != nil {
+		return nil, err
+	}
+	q := req.URL.Query()
+	q.Add("arg", p)
+	if offset != nil {
+		q.Add("offset", fmt.Sprintf("%d", *offset))
+	}
+	if length != nil {
+		q.Add("length", fmt.Sprintf("%d", *length))
+	}
+	req.URL.RawQuery = q.Encode()
+	client := http.DefaultClient
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		if retErr != nil {
+			io.Copy(io.Discard, resp.Body)
+			resp.Body.Close()
+		}
+	}()
+	if resp.StatusCode != 200 {
+		return nil, fmt.Errorf("failed to cat %v; status code: %v", p, resp.StatusCode)
+	}
+	return resp.Body, nil
+}
+
+// Add adds the provided data to IPFS and returns its CID (v1).
+func Add(ipfsAPI string, r io.Reader) (cidv1 string, retErr error) {
+	if ipfsAPI == "" {
+		return "", fmt.Errorf("specify IPFS API address")
+	}
+	ipfsAPIAdd := ipfsAPI + "/api/v0/add"
+	pr, pw := io.Pipe()
+	mw := multipart.NewWriter(pw)
+	contentType := mw.FormDataContentType()
+	go func() {
+		fw, err := mw.CreateFormFile("file", "file")
+		if err != nil {
+			pw.CloseWithError(err)
+			return
+		}
+		if _, err := io.Copy(fw, r); err != nil {
+			pw.CloseWithError(err)
+			return
+		}
+		if err := mw.Close(); err != nil {
+			pw.CloseWithError(err)
+			return
+		}
+		pw.Close()
+	}()
+	req, err := http.NewRequest("POST", ipfsAPIAdd, pr)
+	if err != nil {
+		return "", err
+	}
+	req.Header.Add("Content-Type", contentType)
+	q := req.URL.Query()
+	q.Add("cid-version", "1")
+	q.Add("pin", "true")
+	req.URL.RawQuery = q.Encode()
+	client := http.DefaultClient
+	resp, err := client.Do(req)
+	if err != nil {
+		return "", err
+	}
+	defer func() {
+		io.Copy(io.Discard, resp.Body)
+		resp.Body.Close()
+	}()
+	if resp.StatusCode != 200 {
```

Comment:
Some server may potentially return other 2xx codes?

Reply:
Fixed to allow 2xx.